### PR TITLE
Implement offices API

### PR DESCRIPTION
**Pull request by Windawz**

### CODE REVIEW

File: src/Microservices/Offices/Api/Controllers/OfficesController.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+        return Ok(response);
+    }
+
+    [HttpGet]
+    [ProducesResponseType<ICollection<GetOfficeResponse>>(StatusCodes.Status200OK)]
+    [ProducesResponseType<ICollection<GetOfficeResponse>>(StatusCodes.Status204NoContent)]
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
+    [ProducesResponseType(StatusCodes.Status404NotFound)]
+    public async Task<IActionResult> GetPage(int count, Guid? start)
```
Comment (MikhadziukAliaksei) [inline src/Microservices/Offices/Api/Controllers/OfficesController.cs:40]: If I understand correctly u mean here pagination in this case please use model with skip, take logic

File: src/Microservices/Offices/Api/Controllers/OfficesController.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    {
+        var offices = await _officeService.GetOfficePageAsync(count, start);
+        
+        if (offices.Count > 0)
+        {
+            return Ok(offices);
+        }
+        else
+        {
+            return NoContent();
```
Comment (MikhadziukAliaksei) [inline src/Microservices/Offices/Api/Controllers/OfficesController.cs:50]: If u try to get values u should return empty collection
NoContent it's more about update actions

File: src/Microservices/Offices/Application/Services/Implementations/OfficeService.cs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    public async Task<ICollection<Office>> GetOfficePageAsync(int count, Guid? start = null)
+    {
+        if (start is Guid guid && await _officeRepository.GetAsync(guid) is null)
+        {
+            throw new OfficeNotFoundException(guid);
+        }
+
+        var officeEntities = await _officeRepository.GetPageAsync(count, start);
+
+        // Ugly ToList() call.
```
Comment (MikhadziukAliaksei) [inline src/Microservices/Offices/Application/Services/Implementations/OfficeService.cs:43]: delete unnecessary comments

File: src/Microservices/Offices/Application/Services/Implementations/OfficeService.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        return officeEntity.Id;
+    }
+
+    /// <exception cref="OfficeNotFoundException"/>
+    public async Task EditOfficeAsync(Guid id, OfficeEditInput input)
+    {
+         // Didn't want to use reflection here due to how
+        // slow it is.
+
+        OfficeEntity? officeEntity = await _officeRepository.GetAsync(id);
```
Comment (MikhadziukAliaksei) [inline src/Microservices/Offices/Application/Services/Implementations/OfficeService.cs:72]: U use Patch http method but seems like u updated all model. Look about JsonPath

File: src/Microservices/Offices/Data/Repositories/Implementations/OfficeRepository.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+{
+    public OfficeRepository(OfficesDbContext dbContext) : base(dbContext) { }
+
+    public async Task<ICollection<OfficeEntity>> GetPageAsync(int count, Guid? start = null)
+    {
+            // The drawback of paging at the data access layer
+           // is that we prevent any opportunities for streaming.
+          //
+         // Should've tried IAsyncEnumerable instead, then
+        // implemented pagination at the application/API layer.
```
Comment (MikhadziukAliaksei) [inline src/Microservices/Offices/Data/Repositories/Implementations/OfficeRepository.cs:19]: Use Skip Take logic and then new model for paged rusult

File: src/Microservices/Offices/Data/Migrations/20231222131349_AddOfficeEntity.cs [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        /// <inheritdoc />
+        protected override void Up(MigrationBuilder migrationBuilder)
+        {
+            migrationBuilder.CreateTable(
+                name: "Offices",
+                columns: table => new
+                {
+                    Id = table.Column<Guid>(type: "uuid", nullable: false),
+                    City = table.Column<string>(type: "text", nullable: false),
```
Comment (MikhadziukAliaksei) [inline src/Microservices/Offices/Data/Migrations/20231222131349_AddOfficeEntity.cs:19]: Constraints(max length) for each text row for performance

File: src/Microservices/Offices/Api/Controllers/OfficesController.cs (hunk 1 above)
Comment (Windawz) [reply src/Microservices/Offices/Api/Controllers/OfficesController.cs:40]: Yes, pagination is implied here. If I got it right, you mean switching to offset pagination? If so, [even Microsoft recommends to not use it in favor of keyset pagination](https://learn.microsoft.com/en-us/ef/core/querying/pagination#keyset-pagination). Please, correct me if I'm wrong.

File: src/Microservices/Offices/Application/Services/Implementations/OfficeService.cs (hunk 4 above)
Comment (Windawz) [reply src/Microservices/Offices/Application/Services/Implementations/OfficeService.cs:72]: > Look about JsonPath

I guess you meant Json**Patch**? It seems to handle patching better indeed, I'll try sticking it in. Thanks.

> U use Patch http method but seems like u updated all model.

It doesn't exactly update the whole model. If you take a look at the parameters, you'll notice that one of them is `OfficeEditInput`:

https://github.com/Windawz/InnowiseClinic/blob/ecd6396f488d906316b4d25bb7741399e2e0b8f3/src/Microservices/Offices/Application/Models/OfficeEditInput.cs#L3-L9

As you can see, all of the parameters are nullable and non-required. Same for the DTO that gets mapped to it:

https://github.com/Windawz/InnowiseClinic/blob/ecd6396f488d906316b4d25bb7741399e2e0b8f3/src/Microservices/Offices/Api/DataTransferObjects/Requests/EditOfficeRequest.cs#L3-L11

When a request is sent to the PATCH endpoint of the office controller, it's not expected to have values for all of the properties. In the following snippet, any received values are assigned. If a value hasn't been received for a property, it'll be null:

https://github.com/Windawz/InnowiseClinic/blob/ecd6396f488d906316b4d25bb7741399e2e0b8f3/src/Microservices/Offices/Application/Services/Implementations/OfficeService.cs#L81-L89

File: src/Microservices/Offices/Data/Repositories/Implementations/OfficeRepository.cs (hunk 5 above)
Comment (Windawz) [reply src/Microservices/Offices/Data/Repositories/Implementations/OfficeRepository.cs:19]: I've already commented on the skip/take thing above.

As for the "new model", I suppose it's related to reducing the amount of data sent when requesting a page of offices? US-29 also lists only three fields, so I suppose it's to better fit the requirements too.

If a new model has to be added at the data layer then a whole new entity class would have to be added, due to reliance on EF Core. It'd also need to not have its own table but be derived from `OfficeEntity`. I'll think of something, but no ideas right now.

File: src/Microservices/Offices/Application/Services/Implementations/OfficeService.cs (hunk 3 above)
Comment (Windawz) [reply src/Microservices/Offices/Application/Services/Implementations/OfficeService.cs:43]: Done.

File: src/Microservices/Offices/Data/Migrations/20231222131349_AddOfficeEntity.cs (hunk 6 above)
Comment (Windawz) [reply src/Microservices/Offices/Data/Migrations/20231222131349_AddOfficeEntity.cs:19]: Makes sense. Will do.

File: src/Microservices/Offices/Api/Controllers/OfficesController.cs (hunk 1 above)
Comment (Windawz) [reply src/Microservices/Offices/Api/Controllers/OfficesController.cs:40]: Actually you may be right. The article above mentions that offset pagination should be preferred for random access. I also don't think that there's any reliable ordering when using guids. Will change.

Comment (Windawz) [reply src/Microservices/Offices/Api/Controllers/OfficesController.cs:40]: Apparently PostgreSQL sorts/orders guids by their binary representation using `memcmp`: https://stackoverflow.com/questions/43238882/how-does-ordering-by-uuid-work-in-postgresql

File: src/Microservices/Offices/Application/Services/Implementations/OfficeService.cs (hunk 4 above)
Comment (Windawz) [reply src/Microservices/Offices/Application/Services/Implementations/OfficeService.cs]: Added JsonPatch in b79d6c0.

File: src/Microservices/Offices/Data/Migrations/20231222131349_AddOfficeEntity.cs (hunk 6 above)
Comment (Windawz) [reply src/Microservices/Offices/Data/Migrations/20231222131349_AddOfficeEntity.cs:19]: Fixed by e3d495f.

File: src/Microservices/Offices/Application/Services/Implementations/OfficeService.cs (hunk 3 above)
Comment (Windawz) [reply src/Microservices/Offices/Application/Services/Implementations/OfficeService.cs]: b693077, 1822921.

File: src/Microservices/Offices/Api/Controllers/OfficesController.cs (hunk 2 above)
Comment (Windawz) [reply src/Microservices/Offices/Api/Controllers/OfficesController.cs]: c99d2e6.

File: src/Microservices/Offices/Api/Controllers/OfficesController.cs (hunk 1 above)
Comment (Windawz) [reply src/Microservices/Offices/Api/Controllers/OfficesController.cs]: Replaced with offset pagination in af6c26e.

**Pull request merged**